[AArch64] Disconnect FeatureUseScalarIncVL from FeatureSVE2.

davemgreen opened the pull request:
FeatureUseScalarIncVL is a tuning feature, used to control whether addvl or add+cnt is used. It was previously added as a dependency for FeatureSVE2, an architecture feature but this can be seen as a layering violation. The main disadvantage is that -use-scalar-inc-vl cannot be used without disabling sve2 and all dependant features.

This patch attempt to break that, enabling the feature if SVE2 or SME is present and the use-scalar-inc-vl is not otherwise specified, but not requiring a hard dependency between the two. (Unfortunately I had to check if the value was present in the features string to check that).

It also inverts the meaning of the feature to DontUseScalarIncVL, so that the default state is to use addvl, hopefully the more sensible default going forward.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64Features.td:
```diff
@@ -797,6 +794,9 @@ def FeatureUseFixedOverScalableIfEqualCost: SubtargetFeature<"use-fixed-over-sca
   "UseFixedOverScalableIfEqualCost", "true",
   "Prefer fixed width loop vectorization over scalable if the cost-model assigns equal costs">;
 
+def FeatureDontUseScalarIncVL : SubtargetFeature<"dont-use-scalar-inc-vl",
```
Can you implement this without changing the feature name?

davemgreen replied:
Yeah that should be OK. It should at least make the patch simple and I think just involves inverting all the logic, so turning the option on when we have SVE2/SME (as opposed to turning it off when we have only SVE-1). Having it default to addvl felt like a better default.  Do you have a particular reason to not want to change it?

(I'm not sure if dont-use-scalar-inc-vl is the best name, to be honest. It might be better to have something like preferAddCntToAddvl to avoid the negatives and make it a little more explicit)

sdesmalen-arm replied:
I found the double negative in `-mattr=-dont-use-scalar-inc-vl` confusing and the current logic is a bit simpler (although I'm not particularly happy with the explicit partial string matching, which I guess is the downside of doing it this way). A name change would avoid the double negative, but thinking about this; why don't we just make it an LLVM option rather than a feature? 

That way, we can do something like:
```
if (ForceScalarIncVL.getNumOccurrences())
  UseScalarIncVL = ForceScalarIncVL.getValue();
else if (hasSVE2() || hasSME())
  UseScalarIncVL = true;
// else UseScalarIncVL = false; (would be default)
```

We can still make use of the `UseScalarIncVL`  Predicate in TableGen, it just wouldn't be defined through a target feature.

davemgreen replied:
Yeah the string matching isn't the greatest. We only do it once though, and I believe this is the only way to check if a feature was specified. I assume that it was made a target feature so that we could add it to certain cpus if it is useful. I'm not 100% sure we won't find that would still be useful in some cases, but we have progressed a bit in time now and it hasn't come up yet.

davemgreen committed:
[AArch64] Remove FeatureUseScalarIncVL

FeatureUseScalarIncVL is a tuning feature, used to control whether addvl or
add+cnt is used. It was previously added as a dependency for FeatureSVE2, an
architecture feature but this can be seen as a layering violation. The main
disadvantage is that -use-scalar-inc-vl cannot be used without disabling sve2
and all dependant features.

This patch now replaces that with an option that if unset defaults to hasSVE ||
hasSME, but is otherwise overriden by the option. The hope is that no cpus will
rely on the tuning feature (or we can readdit if needed.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64Subtarget.h:
```diff
@@ -417,6 +417,8 @@ class AArch64Subtarget final : public AArch64GenSubtargetInfo {
     return DefaultSVETFOpts;
   }
 
+  bool useScalarIncVL() const;
```
nit: maybe add a comment to explain what `useScalarIncVL` means?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64Subtarget.cpp:
```diff
@@ -575,6 +579,14 @@ void AArch64Subtarget::mirFileLoaded(MachineFunction &MF) const {
+  // If SVE2 or SME is present (we are not SVE-1 only) and UseScalarIncVL
+  // is not otherwise set, enable it by default.
+  if (UseScalarIncVL.getNumOccurrences())
+    return UseScalarIncVL;
```
I'd still recommend creating a separate variable for it in the Subtarget that gets initialised with the code you have here. That way a CPU could possibly override it in `AArch64Subtarget::initializeProperties` (if there is ever a need for that) and then the call to `useScalarIncVL()` can be inlined rather than ending up as a call. But this is just me nitpicking a bit :)

davemgreen replied:
Thanks - We would probably want go the other way and reintroduce the target features if we needed it for a specific CPU. The preferred way is for tuning features to control the options, which are applied to given cpus that they are relevant to.

davemgreen commented:
This was 1d6d073fbbaebbde6891501fe20f02a0ea345131.
